### PR TITLE
PEP 694: Upload 2.0 API for Python Package Repositories

### PR DESCRIPTION
**Pull request by dstufft**

### CODE REVIEW

**Comment by AA-Turner** [comment]
You can use 694.

A

**Comment by AA-Turner** [comment]
I don't have time for a full review at the moment, but should this have `Topic: Packaging`?

A

**Comment by dstufft** [comment]
I’ll fix that it should yes. I wrote it originally prior to that tag existing :)

File: pep-0694.rst [hunk 1]
```diff
@@ -0,0 +1,692 @@
...
+=============
+
+This PEP traces the root cause of most of the issues with the existing API to be
+roughly two things:
+
+- The metadata is submitted alongside the file, rather than being parsed from the
+  file itself.
+
+  - This is actually fine if used as a pre-check, but it should be reconciled
+    against the actual ``METADATA`` or similiar files within the distribution.
```
[the remark this answers is not in the document]
Comment (dstufft) [reply pep-0694.rst:121]: I'm bad at words ok.

File: pep-0694.rst [hunk 2]
```diff
@@ -0,0 +1,731 @@
...
+Already uploaded files may be deleted by issuing a ``DELETE`` request to the file
+upload URL without the ``Upload-Token``.
+
+A successful deletion request **MUST** response with a ``204 No Content``.
+
+
+Session Status
+~~~~~~~~~~~~~~
+
+Similiarly to file upload, the session URL is provided in the response to
```
Comment (JelleZijlstra) [inline pep-0694.rst:444]:
```suggestion
Similarly to file upload, the session URL is provided in the response to
```

(I got distracted halfway through, I'll read the rest now.)

File: pep-0694.rst [hunk 3]
```diff
@@ -0,0 +1,731 @@
...
+~~~~~~~~~~~~~~~~~~~
+
+To cancel an upload session, a client issues a ``DELETE`` request to the
+same session URL as before. At which point the server marks the session as
+canceled, **MAY** purge any data that was uploaded as part of that session,
+and future attempts to access that session URL or any of the file upload URLs
+**MAY** return a ``404 Not Found``.
+
+To prevent a lot of dangling sessions, servers may also choose to cancel a
+session on it's own accord. It is recommended that servers expunge their
```
Comment (JelleZijlstra) [inline pep-0694.rst:464]:
```suggestion
session on their own accord. It is recommended that servers expunge their
```

Match plural "servers"

File: pep-0694.rst [hunk 4]
```diff
@@ -0,0 +1,731 @@
...
+~~~~~~~~~~~~~~~~~~
+
+To complete a session, and publish the files that have been included in it,
+a client **MUST** send a ``POST`` request to the ``publish`` url in the
+session status payload.
+
+If the server is able to immediately complete the session, it may do so
+and return a ``201 Created`` response. If it is unable to immediately
+complete the session (for instance, if it needs to do processing that may
+take longer than reasonable in a single http request), then it may return
```
Comment (JelleZijlstra) [inline pep-0694.rst:479]:
```suggestion
take longer than reasonable in a single HTTP request), then it may return
```

File: pep-0694.rst [hunk 5]
```diff
@@ -0,0 +1,731 @@
...
+This has one big benefit, that if a client doesn't care about resuming their
+download, the work to support, from a client side, resumable uploads is able
+to be completely ignored. They can just ``POST`` the file to the URL, and if
+it doesn't succeed, they can just ``POST`` the whole file again.
+
+The other benefit is that even if you do want to support resumption, you can
+still just ``POST`` the file, and unless you *need* to resume the download,
+that's all you have to do.
+
+Another, possibly theortical, benefit is that for hashing the uploaded files,
```
Comment (JelleZijlstra) [inline pep-0694.rst:657]:
```suggestion
Another, possibly theoretical, benefit is that for hashing the uploaded files,
```

File: pep-0694.rst [hunk 6]
```diff
@@ -0,0 +1,731 @@
...
+the serial chunks requirement means that the server can maintain hashing state
+between requests, update it for each request, then write that file back to
+storage. Unfortunately this isn't actually possible to do with Python's hashlib,
+though there is some libraries like `Rehash <https://github.com/kislyuk/rehash>`_
```
Comment (JelleZijlstra) [inline pep-0694.rst:661]:
```suggestion
though there are some libraries like `Rehash <https://github.com/kislyuk/rehash>`_
```

File: pep-0694.rst [hunk 7]
```diff
@@ -0,0 +1,731 @@
...
+that implement it, but they don't support every hash that hashlib does
+(specifically not blake2 or sha3 at the time of writing).
+
+We might also need to reconstitute the download for processing anyways to do
+things like extract metadata, etc from it, which would make it a moot point.
+
+The downside is that there is no ability to parallelize the upload of a single
+file because each chunk has to be submitted serially.
+
+AWS S3 has a similiar API (and most blob stores have copied it either wholesale
```
Comment (JelleZijlstra) [inline pep-0694.rst:671]:
```suggestion
AWS S3 has a similar API (and most blob stores have copied it either wholesale
```

File: pep-0694.rst [hunk 8]
```diff
@@ -0,0 +1,731 @@
...
+specific parts.
+
+This has a big benefit in that it allows parallelization in uploading files,
+allowing clients to maximize their bandwidth using multiple threads to send
+the data.
+
+We wouldn't need an explicit step (1), because our session would implicitly
+initiate a multipart upload for each file.
+
+It does have it's own downsides:
```
Comment (JelleZijlstra) [inline pep-0694.rst:693]:
```suggestion
It does have its own downsides:
```

File: pep-0694.rst [hunk 9]
```diff
@@ -0,0 +1,731 @@
...
+
+- Clients have to do more work on every request to have something resembling
+  resumble uploads. They would *have* to break the file up into multiple parts
```
Comment (JelleZijlstra) [inline pep-0694.rst:696]:
```suggestion
  resumable uploads. They would *have* to break the file up into multiple parts
```

File: pep-0694.rst [hunk 10]
```diff
@@ -0,0 +1,731 @@
...
+  single part.
+
+  - S3 works around this by having another API for one shot uploads, but
+    I'd rather not have two different APIs for uploading the same file.
+
+- Verifying hashes gets somewhat more complicated. AWS implements hashing
+  multipart uploads by hashing each part, then the overall hash is just a
+  hash of those hashes, not of the content itself. We need to know the
+  actual hash of the file itself for PyPI, so we would have to reconstitute
+  the file and read it's content and hash it once it's been fully uploaded,
```
Comment (JelleZijlstra) [inline pep-0694.rst:711]:
```suggestion
  the file and read its content and hash it once it's been fully uploaded,
```

**Comment by dstufft** [comment]
Oops sorry, I'll send a follow up PR that fixes those.

**Pull request merged**